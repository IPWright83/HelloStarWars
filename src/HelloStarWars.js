import React, { Component } from "react";

class HelloStarWars extends React.PureComponent {

	constructor(props) {
		super(props);
		this.state = {};
	}

	async componentDidMount() {
		const id = this.props.id;

		try {
			const response = await fetch(`https://swapi.co/api/people/${id}`);
			const { name } = await response.json();
			this.setState({ name });
		} catch (e) {
			this.setState({ name: "!dlroW... Oh dear, something went wrong" });
		}
	}

	render() {
		const name = this.state.name || "...";
		return(
			<h1>Hello {name}!</h1>
		);
	}
}

export default HelloStarWars;