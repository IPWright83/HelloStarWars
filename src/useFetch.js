import { useState, useEffect } from "react";

export default function useFetch(url) {
  const [data, setData] = useState({});

  async function getData() {
    try {
      const response = await fetch(url);
      const data = await response.json();
      setData(data);
    } catch (err) {
      setData({ name: "!dlroW... Oh dear, something went wrong" });
    }
  }

  useEffect(
    () => {
      getData();
    },
    [url]
  );

  return data;
}
