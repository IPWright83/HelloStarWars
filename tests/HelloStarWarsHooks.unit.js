import React from "react";
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';

import HelloStarWars from "../src/HelloStarWarsHooks";
import useFetch from "../src/useFetch";

//const mockFetch = jest.fn();
//jest.mock("../src/useFetch", () => () => mockFetch());

describe("HelloStarWarsHooks", () => {

	it("Renders with a place holder", () => {
		// Render our component
		const { asFragment } = render(<HelloStarWars id="1" />);

		expect(asFragment()).toMatchSnapshot();
	});

	describe("ComponentDidMount", () => {
		
		beforeEach(() => {
			fetch.resetMocks();
		});

		it("Obtains result and sets data correctly", async () => {
			let result;

			fetch.mockResponseOnce(JSON.stringify({ name: 'Jedi Master' }));

			// Render our component within an act to flush the useFetch
			act(() => {
				const { asFragment } = render(<HelloStarWars id="1" />);
				result = asFragment();
			});

			expect(result).toMatchSnapshot();
		});

		it("Handles an error in the web request correctly", async () => {
			let result;

			fetch.mockReject(new Error("Invalid ID"));

			// Render our component within an act to flush the useFetch
			act(() => {
				const { asFragment } = render(<HelloStarWars id="foobar" />);
				result = asFragment();
			});
			
			expect(result).toMatchSnapshot();
		});
	});
});