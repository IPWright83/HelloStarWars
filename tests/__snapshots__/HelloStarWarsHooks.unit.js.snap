// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloStarWarsHooks ComponentDidMount Handles an error in the web request correctly 1`] = `
<DocumentFragment>
  <h1>
    Hello !dlroW... Oh dear, something went wrong!
  </h1>
</DocumentFragment>
`;

exports[`HelloStarWarsHooks ComponentDidMount Obtains result and sets data correctly 1`] = `
<DocumentFragment>
  <h1>
    Hello Jedi Master!
  </h1>
</DocumentFragment>
`;

exports[`HelloStarWarsHooks Renders with a place holder 1`] = `
<DocumentFragment>
  <h1>
    Hello ...!
  </h1>
</DocumentFragment>
`;
